### PR TITLE
Unify method-source representation — retire the flush reindent reconciliation (BT-2584)

### PR DESCRIPTION
**Pull request by jamesc**
Linear: https://linear.app/beamtalk/issue/BT-2584

## Summary

BT-2577 fixed flush corrupting doc-commented methods by *reconciling* two different shapes of "a method's source" at flush time (`beamtalk_workspace_flush:reindent/2`). This PR eliminates the mismatch instead: the stored ChangeLog `source` is now the on-disk byte-span shape (`source_ref == disk[span]`) **by construction**, so `Workspace flush` splices it verbatim and the whole shape-reconciliation class of bugs is retired.

The single sanctioned reshape lives in Rust (`unparse::reindent_method_source`) and is applied once, at the live-patch install hook, where both the canonical body and the disk slice (base indentation + trailing-newline state) are known.

## Key changes

- **`crates/beamtalk-core/src/unparse/mod.rs`** — new `reindent_method_source(base_indent, source)`: strips a column-0 body's shared indent, re-prepends the base, keeps blank lines empty. The one place the reshape is defined.
- **`crates/beamtalk-compiler-port/src/main.rs`** — new `reindent_method_source` port command; FFI wrappers in `beamtalk_compiler` / `beamtalk_compiler_server` / `beamtalk_compiler_port`.
- **`beamtalk_repl_loader.erl`** (`store_disk_shaped_entry/4`) — at method-patch install, reshapes the compiler's canonical column-0 body to the span's base indentation (via the FFI) and matches the disk slice's trailing-newline state, storing the disk-shaped body. Falls back to non-flushable (memory-only) if the FFI is unavailable, so a column-0 body can never reach flush.
- **`beamtalk_workspace_flush.erl`** — `splice_with_prev` now splices the stored body verbatim; `reindent/2` and its helpers removed.

## Tests

- Rust unit tests for `reindent_method_source` covering the trivia edge cases the issue called out: blank line between doc and method, inline `//` comments, multi-line signature, tab indentation, no-trailing-newline, idempotence.
- Compiler-port command tests.
- Corpus round-trip tests (extend the existing `method_span` corpus):
  - `corpus_reshape_round_trip_is_byte_identical` — the reindent/dedent + trailing-newline algebra is the identity on every method's verbatim disk slice.
  - `corpus_canonical_methods_round_trip_byte_identical` — the **full** `parse_method → unparse_method → reindent` pipeline reproduces `disk[span]` byte-for-byte for every already-canonical method (the in-scope "unchanged save → byte-identical flush" guarantee, isolated from the unparser's separate formatting fidelity).
- BT-2577 corruption-regression flush test updated to the disk-shaped stored body (no regression).

## Acceptance criteria

- [x] Single documented representation; `source_ref == disk[span]` by construction.
- [x] `reindent/2` removed; the splice needs no reshaping.
- [x] Round-trip property holds across the stdlib + examples corpus.
- [x] Trivia edge cases covered (blank line, inline comments, multi-line signature, tabs).
- [x] No regression in BT-2577's corruption tests.

## Notes / follow-ups (pre-existing, out of scope)

- The `selector_not_found` append path (a brand-new method added live with no disk span) still appends the canonical column-0 body un-indented — unchanged by this PR. Worth a follow-up to unify the append path with the same reshape.
- `browse-method-source`'s live `__source__` and `disk_differs` remain column-0 (a codegen concern touching SystemNavigation scanners); not regressed here.
- The full-pipeline byte-identity rate across the *whole* corpus is ~83% — the remainder is the unparser reformatting hand-written methods (pre-existing fidelity, not this change). The canonical-subset test pins the BT-2584 guarantee at 100% for that subset.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Aw8Wg2gsyXCAy5hXHdNvsL

---
_Generated by [Claude Code](https://claude.ai/code/session_01Aw8Wg2gsyXCAy5hXHdNvsL)_

### CODE REVIEW

File: runtime/apps/beamtalk_workspace/src/beamtalk_repl_loader.erl [hunk 1]
```diff
@@ -1448,17 +1448,106 @@ add_span_or_downgrade(Base, ClassNameBin, SelectorBin, Side, SourceFile, AbsPath
...
+%% construction (BT-2584). A later `Workspace flush' then splices it verbatim
+%% with no reindent. The base indentation is the leading whitespace of the disk
+%% slice (`PrevSource') the patch replaces. If the reshape FFI is unavailable,
+%% downgrade to memory-only rather than store a column-0 body that flush would
+%% splice into an indented region and corrupt the file.
+-spec store_disk_shaped_entry(
+    map(), binary(), #{start := non_neg_integer(), 'end' := non_neg_integer()}, binary()
+) -> map().
+store_disk_shaped_entry(Base, SourceFile, Span, PrevSource) ->
+    Canonical = maps:get(source, Base, <<>>),
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_workspace/src/beamtalk_repl_loader.erl]: Fixed in b694197. Switched to a strict head match `store_disk_shaped_entry(#{source := Canonical} = Base, ...)`. Verified `source` is always populated on the only caller path (`do_emit_change_entry` builds `Base` with `source => Source`), so this can only crash on a genuine caller-contract violation — exactly the let-it-fail behavior you want, rather than silently storing a method-destroying body. 253 workspace EUnit tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/apps/beamtalk_workspace/src/beamtalk_repl_loader.erl [hunk 2]
```diff
@@ -1448,17 +1448,106 @@ add_span_or_downgrade(Base, ClassNameBin, SelectorBin, Side, SourceFile, AbsPath
...
+    BaseIndent = leading_ws(PrevSource),
+    case beamtalk_compiler:reindent_method_source(Canonical, BaseIndent) of
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_workspace/src/beamtalk_repl_loader.erl:1472]: Good architectural point — this is a genuine tradeoff (single Rust source of truth vs. a transport-level failure path on the install hook that the pure-Erlang `reindent/2` didn't have). It's a deliberate design choice for BT-2584 rather than something to flip in this PR, so I've tracked it as BT-2592 — evaluate a pure-Erlang mirror (held byte-identical to the Rust impl by the corpus round-trip tests) to drop the failure path, and fold in the duplicated `strip_trailing_newlines`/`ensure_trailing_newline`/`leading_ws` helpers nit at the same time: https://linear.app/beamtalk/issue/BT-2592

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/apps/beamtalk_compiler/src/beamtalk_compiler_port.erl [hunk 3]
```diff
@@ -863,6 +864,73 @@ handle_method_span_response(Other) ->
...
+                end),
+                {error, port_error, <<"Compiler port timed out">>}
+            end
+    catch
+        error:badarg ->
+            ?LOG_ERROR("Compiler port not available (reindent)", #{
+                domain => [beamtalk, runtime], port => Port
+            }),
+            {error, port_error, <<"Compiler port is not available">>}
+    end.
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_compiler/src/beamtalk_compiler_port.erl]: Fixed in 3b267b8. Added the wildcard fallback. I used `{error, bad_argument, _}` rather than `port_error` to match `resolve_method_span/5` exactly — it's a bad-argument condition, not a transport failure — and widened the spec to `{error, port_error | bad_argument, binary()}`. The caller's `{error, Reason, _}` branch handles it (downgrade to memory-only).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/apps/beamtalk_workspace/src/beamtalk_repl_loader.erl [hunk 4]
```diff
@@ -1448,17 +1448,105 @@ add_span_or_downgrade(Base, ClassNameBin, SelectorBin, Side, SourceFile, AbsPath
...
+    end.
+
+%% The leading run of spaces/tabs of the first line of `Body' — the base
+%% indentation of the on-disk method definition the span covers.
+-spec leading_ws(binary()) -> binary().
+leading_ws(Body) ->
+    First =
+        case binary:split(Body, <<"\n">>) of
+            [Line | _] -> Line;
+            [] -> <<>>
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_workspace/src/beamtalk_repl_loader.erl]: Fixed in 3b267b8. Dropped the unreachable `[]` arm and simplified to `[First | _] = binary:split(Body, <<"\n">>)` with a comment noting `binary:split/2` always returns a non-empty list.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/apps/beamtalk_compiler/src/beamtalk_compiler_port.erl [hunk 5]
```diff
@@ -863,6 +864,75 @@ handle_method_span_response(Other) ->
...
+                                domain => [beamtalk, runtime], port => Port
+                            }),
+                            {error, port_error, <<"Compiler port response is malformed">>}
+                    end;
+                {Port, {exit_status, Status}} ->
+                    ?LOG_ERROR("Compiler port exited during reindent query", #{
+                        domain => [beamtalk, runtime], status => Status
+                    }),
+                    {error, port_error, <<"Compiler port exited unexpectedly">>}
+            after 30000 ->
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_compiler/src/beamtalk_compiler_port.erl]: Fixed in 4618520. Dropped the reindent receive timeout to 5000 ms, with a comment noting why it differs from the 30s parse/search commands (pure in-memory reshape → a stall is a wedged port, not slow work; fail fast to limit blast radius on the shared gen_server).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/apps/beamtalk_compiler/src/beamtalk_compiler_port.erl [hunk 6]
```diff
@@ -863,6 +864,75 @@ handle_method_span_response(Other) ->
...
+                {error, port_error, <<"Compiler port timed out">>}
+            end
+    catch
+        error:badarg ->
+            ?LOG_ERROR("Compiler port not available (reindent)", #{
+                domain => [beamtalk, runtime], port => Port
+            }),
+            {error, port_error, <<"Compiler port is not available">>}
+    end;
+reindent_method_source(_Port, _Source, _BaseIndent) ->
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_compiler/src/beamtalk_compiler_port.erl:928]: Agreed — that's exactly the intent: a defensive net for direct port-module callers, since the public `beamtalk_compiler_server` entry point is typed `(binary(), binary())` and dialyzer catches mis-typed callers at compile time. No integration test needed; the clause just keeps the two sibling port commands consistent. Leaving as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_
